flow/layers/layer.h

```diff
@@ -171,10 +186,26 @@ class Layer {
     paint_bounds_ = paint_bounds;
   }
 
+  void set_paint_region(const PaintRegion& paint_region) {
+    paint_region_ = paint_region;
```

Comment:
Storing this structure in the layer becomes problematic when dealing with retained layers. If this layer == old_layer then we are overwriting a field that contained useful information from the time when the old layer tree was diffed.

Theoretically this feels like an accident waiting to happen.
In practice, it may be that when we have retained layers, they are never in conflict and so we would not need the paint region of the previous layer (to be added to a damage region). But, what if a transform above the retained layer changed? That would lead to the same layer having different regions in the two scenes, but they may be the same physical layer.

Is there a simple way to have this region be "per layer per DiffContext"?

Reply:
If the transform above the retained layer is changed, then the paint_region of retained layer is never read.

```cpp
TransformLayer::Diff(...) {
   ...
    if (transform_ != prev->transform_) {
      context->MarkSubtreeDirty(old_layer->paint_region());
    }
   ...
}
```
So the region of old transform layer above is marked as dirty. After this, no more diffing is done in the subtree, everything is diffed as "new", so the paint region in retained layer will never be read, it will only be written to.

Reply:
Right. I see a problem here if rasterization gets cancelled during preroll.

Reply:
One potential solution for this would be to store the transform matrix together with paint region. Then if we find out that the matrix changed between frames, we can convert the paint region to new matrix.

Reply:
[This commit](https://github.com/flutter/engine/pull/21498/commits/f8d47c311fd3b3f23fb800b0aa4890837c12b879) moves the paint region from layer to layer_tree. Which makes it paint region per layer per layer_tree. So when diffing layer trees, we get correct regions for retained layers regardless of possible discarded intermediate layer trees.